doc/stdenv/stdenv.chapter.md

```diff
@@ -790,6 +790,14 @@ There’s many more kinds of arguments, they are documented in `nixpkgs/pkgs/bui
 
 `wrapProgram` is a convenience function you probably want to use most of the time.
 
+### `makeBinaryWrapper` \<executable\> \<wrapperfile\> \<args\> {#fun-makeBinaryWrapper}
+
+A setup-hook very similar to `makeWrapper`, only it creates a tiny _compiled_ wrapper executable, that can be used as a shebang interpreter. This is needed mostly on Darwin, where shebangs cannot point to scripts, [due to a limitation with the `execve`-syscall](https://stackoverflow.com/questions/67100831/macos-shebang-with-absolute-path-not-working). The arguments it accepts are <!-- TODO: make them identical? --> similar to those of `makeWrapper` and they are documented in `nixpkgs/pkgs/build-support/setup-hooks/make-binary-wrapper.sh`.
```

Comment:
Yeah, `system(...)` won't be able to replace `--run "cd som-dir"`, since using system is more analogous to wrapping the command in a subshell, while with the old `--run` things are executed at the "top level". Might make sense to add a flag [`--chdir DIR`](https://man7.org/linux/man-pages/man2/chdir.2.html) which replaces the usecase you describe.

Reply:
I'm a bit conflicted between `--set-workdir DIR` and `--chdir DIR` now. What do you think?

Some examples of `--run` in nixpkgs which would be hard to recreate in makeBinaryWrapper:

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/development/libraries/gstreamer/core/default.nix#L99 (changes environment inside --run)

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/tools/graphics/gnuplot/default.nix#L60
(sources a script to change the environment inside --run)

https://github.com/NixOS/nixpkgs/blob/971b383a28f4baee8ea3931af4840fa221929fd6/pkgs/tools/text/popfile/default.nix#L45
(changes environment inside --run in a way which depends on other environment variables)

Reply:
> I'm a bit conflicted between --set-workdir DIR and --chdir DIR now. What do you think?

would go for `--chdir`

> Some examples of --run in nixpkgs which would be hard to recreate in makeBinaryWrapper:

not supporting those yet seems fine to me. if it turns out that it's necessary for something it'd be easy enough to add eg `--env-from-script <script>` that shall output a nul-separated list of `ENV=val` pairs the wrapper then sets in the environment. arguably all the missing actions of makeWrapper (except maybe the ability to chdir) fall in that same category of "not needed yet"

Reply:
According to what I see in `git grep -A3 '\(makeWrapper\|wrapProgram\)' | grep -- --run` the vast majority of usages for `--run` is for `cd` and for making the wrappers add paths for env vars relative to `$HOME`. How about adding a flag: `--set-relative` and `--set-relative-default`? Perhaps also we can add flags such as `--run-c` which will add c code to be embedded?